README.md

A composable, Future-based library for making HTTP requests to GebetaMap.

This package contains a set of high-level functions that make it
easy to consume HTTP resources from GebetaMap. It's multi-platform, and supports mobile, desktop,
and the browser.

## Using
making direction request to gebeta map
```dart
import 'package:gebetamap/gebetamap.dart';
    GebetaMapRequest gmr = new GebetaMapRequest();
     var directionStart = new Map();
     directionStart["lat"] = 9.022528936095531;
    directionStart["lon"] = 38.80400061607361;
    var directionStop = new Map();
    directionStop["lat"] = 9.021755421022991;
     directionStop["lon"] = 38.79883468151093;
    String apiKey = "";
    ResponseData rds = await gmr.direction(directionStart, directionStop, apiKey);

    print('Response message: ${rds.message}');
    print('Response status: ${rds.status}');
    print('Response body: ${rds.path}');

```

making travels salesman request to gebeta map

```dart
import 'package:gebetamap/gebetamap.dart';
var points = <Map>[];
  String apiKey = "";
  points.add({'lat': 9.005980058445639, 'lon': 38.785734616513466});
  points.add({'lat': 9.01166345564756, 'lon': 38.789008246478424});
  GebetaMapRequest gmr = new GebetaMapRequest();
  ResponseData rds = await gmr.tss(points, apiKey);
  print('Response message: ${rds.message}');
  print('Response status: ${rds.status}');
  print('path: ${rds.path}');
  print('distance: ${rds.totalDistance}');
```

making one to many request 
```dart
  var directionStart = new Map();
  directionStart["lat"] = 9.022528936095531;
  directionStart["lon"] = 38.80400061607361;
  var points = <Map>[];
  String apiKey = "";
  points.add({'lat': 9.005980058445639, 'lon': 38.785734616513466});
  points.add({'lat': 9.01166345564756, 'lon': 38.789008246478424});
  GebetaMapRequest gmr = new GebetaMapRequest();
  List<ResponseData> rs = await gmr.OneToMany(directionStart, points, apiKey);

  for (int i = 0; i < rs.length; i++) {
    print(rs[i].message);
    print(rs[i].status);
    print(rs[i].totalDistance);
    print(rs[i].path);
    print("(------------------------------------------)");
  }
```



making Geocoding request
``` dart
    String name = "kotebe 02";
    String apiKey = "";
    GebetaMapRequest gmr = new GebetaMapRequest();
    ResponseData rs = await gmr.geocode(name, apiKey);
    print(rs.data);
```




CHANGELOG.md
## 1.1.3

## 1.1.3

* One to Many direction request added.
* Geocoding added


test/gebetamap_test.dart
import 'package:gebetamap/gebetamap.dart';
import 'package:test/test.dart';

void main() {}


lib/src/gebetamap_base.dart
import 'package:gebetamap/src/utils/constants.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

import 'utils/Resonse.dart';

class GebetaMapRequest {
  Future<ResponseData> direction(
      Map directoinStart, Map directionStop, String apiKey) async {
    try {
      var url = Constants.url +
          "/direction/?la1=" +
          directoinStart["lat"].toString() +
          "&lo1=" +
          directoinStart["lon"].toString() +
          "&la2=" +
          directionStop["lat"].toString() +
          "&lo2=" +
          directionStop["lon"].toString() +
          "&apiKey=" +
          apiKey;
      final response = await http.get(Uri.parse(url));

      var decodedJson = json.decode(response.body);

      if (response.statusCode == 200) {
        var path = decodedJson["direction"] as List;
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString(),
            path: path,
            totalDist: decodedJson["totalDistance"]);
        return rs;
      } else if (response.statusCode == 403) {
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString());
        return rs;
      } else if (response.statusCode == 400) {
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString());
        return rs;
      } else if (response.statusCode == 500) {
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString());
        return rs;
      } else {
        ResponseData rs = ResponseData(messages: "UnKnown Error");
        return rs;
      }
    } catch (err) {
      throw err;
    }
  }

  Future<List<ResponseData>> matrix(var points, String apiKey) async {
    try {
      String _points = '';
      for (int i = 0; i < points.length; i++) {
        if (i == points.length - 1) {
          _points +=
              points[i]['lat'].toString() + "/" + points[i]['lon'].toString();
        } else {
          _points += points[i]['lat'].toString() +
              "/" +
              points[i]['lon'].toString() +
              ",";
        }
      }

      var url =
          Constants.url + "/matrix/?start=" + _points + "&apiKey=" + apiKey;
      final response = await http.get(Uri.parse(url));

      var decodedJson = json.decode(response.body);
      List<ResponseData> responseList = [];
      if (response.statusCode == 200) {
        var decodedResposne = decodedJson["response"] as List;

        for (int i = 0; i < decodedResposne.length; i++) {
          // var path = response[i]["direction"] as List;
          ResponseData rs = ResponseData(
              messages: decodedResposne[i]["data"]["msg"],
              statuss: response.statusCode.toString(),
              totalDist: decodedResposne[i]["data"]["totalDistance"],
              path: decodedResposne[i]["data"]["direction"]);

          responseList.add(rs);
        }
      } else if (response.statusCode == 403) {
        ResponseData rs = ResponseData(
            messages: "Invalid Api Key",
            statuss: response.statusCode.toString());
        responseList.add(rs);
      } else if (response.statusCode == 400) {
        ResponseData rs = ResponseData(
            messages: "Route Not Found",
            statuss: response.statusCode.toString());
        responseList.add(rs);
      } else if (response.statusCode == 500) {
        ResponseData rs = ResponseData(
            messages: "Internal Server Error",
            statuss: response.statusCode.toString());
        responseList.add(rs);
      } else {
        ResponseData rs = new ResponseData(messages: "UnKnown Error");
        responseList.add(rs);
      }
      return responseList;
    } catch (err) {
      throw err;
    }
    //loop through the points and it
  }

// ignore: non_constant_identifier_names
  Future<ResponseData> tss(var points, String apiKey) async {
    try {
      String _points = '';
      for (int i = 0; i < points.length; i++) {
        if (i == points.length - 1) {
          _points +=
              points[i]['lat'].toString() + "/" + points[i]['lon'].toString();
        } else {
          _points += points[i]['lat'].toString() +
              "/" +
              points[i]['lon'].toString() +
              ",";
        }
      }

      var url = Constants.url + "/tss/?start=" + _points + "&apiKey=" + apiKey;
      final response = await http.get(Uri.parse(url));
      var decodedJson = json.decode(response.body);

      if (response.statusCode == 200) {
        var path = decodedJson["direction"] as List;
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString(),
            path: path,
            totalDist: decodedJson["totalDistance"]);
        return rs;
      } else if (response.statusCode == 403) {
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString());
        return rs;
      } else if (response.statusCode == 400) {
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString());
        return rs;
      } else if (response.statusCode == 500) {
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString());
        return rs;
      } else {
        ResponseData rs = ResponseData(messages: "UnKnown Error");
        return rs;
      }
    } catch (err) {
      throw err;
    }
  }

// ignore: non_constant_identifier_names
  Future<List<ResponseData>> OneToMany(
      var directionStart, var points, String apiKey) async {
    try {
      String _points = '';
      for (int i = 0; i < points.length; i++) {
        if (i == points.length - 1) {
          _points +=
              points[i]['lat'].toString() + "/" + points[i]['lon'].toString();
        } else {
          _points += points[i]['lat'].toString() +
              "/" +
              points[i]['lon'].toString() +
              ",";
        }
      }
      var url = Constants.url +
          "/onm/?la1=" +
          directionStart["lat"].toString() +
          "&lo1=" +
          directionStart["lon"].toString() +
          "&json=" +
          _points +
          "&apiKey=" +
          apiKey;

      final response = await http.get(Uri.parse(url));
      // print('Response status: ${response.statusCode}');
      var decodedJson = json.decode(response.body);
      List<ResponseData> responseList = [];
      if (response.statusCode == 200) {
        var decodedResposne = decodedJson["directions"] as List;
        for (int i = 0; i < decodedResposne.length; i++) {
          // var path = response[i]["direction"] as List;
          ResponseData rs = ResponseData(
              messages: decodedResposne[i]["msg"],
              statuss: response.statusCode.toString(),
              totalDist: decodedResposne[i]["totalDistance"],
              path: decodedResposne[i]["direction"]);

          responseList.add(rs);
        }
      } else if (response.statusCode == 403) {
        ResponseData rs = ResponseData(
            messages: "Invalid Api Key",
            statuss: response.statusCode.toString());
        responseList.add(rs);
      } else if (response.statusCode == 400) {
        ResponseData rs = ResponseData(
            messages: "Route Not Found",
            statuss: response.statusCode.toString());
        responseList.add(rs);
      } else if (response.statusCode == 500) {
        ResponseData rs = ResponseData(
            messages: "Internal Server Error",
            statuss: response.statusCode.toString());
        responseList.add(rs);
      } else {
        ResponseData rs = ResponseData(messages: "UnKnown Error");
        responseList.add(rs);
      }
      return responseList;
    } catch (err) {
      throw err;
    }
  }

  Future<ResponseData> geocode(String name, String apiKey) async {
    try {
      var url = "https://mapapi.gebeta.app/api/v1/route/geocoding?name=" +
          name +
          "&apiKey=" +
          apiKey;
      final response = await http.get(Uri.parse(url));
      var decodedJson = json.decode(response.body);
      print(decodedJson);
      if (response.statusCode == 200) {
        var path = decodedJson["data"] as List;
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString(),
            responseData: decodedJson["data"]);
        return rs;
      } else if (response.statusCode == 403) {
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString());
        return rs;
      } else if (response.statusCode == 400) {
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString());
        return rs;
      } else if (response.statusCode == 500) {
        ResponseData rs = ResponseData(
            messages: decodedJson["msg"],
            statuss: response.statusCode.toString());
        return rs;
      } else {
        ResponseData rs = ResponseData(messages: "UnKnown Error");
        return rs;
      }
    } catch (err) {
      throw err;
    }
  }
}


example/gebetamap_example.dart
import 'package:gebetamap/gebetamap.dart';

void callDirection() async {
  GebetaMapRequest gmr = new GebetaMapRequest();
  var directionStart = new Map();
  directionStart["lat"] = 9.022528936095531;
  directionStart["lon"] = 38.80400061607361;
  var directionStop = new Map();
  directionStop["lat"] = 9.021755421022991;
  directionStop["lon"] = 38.79883468151093;
  String apiKey = "";
  ResponseData rds = await gmr.direction(directionStart, directionStop, apiKey);
  print('Response message: ${rds.message}');
  print('Response status: ${rds.status}');
  print('path: ${rds.path}');
  print('distance: ${rds.totalDistance}');
}

void callMatrix() async {
  var points = <Map>[];
  String apiKey = "";
  points.add({'lat': 9.005980058445639, 'lon': 38.785734616513466});
  points.add({'lat': 9.01166345564756, 'lon': 38.789008246478424});
  GebetaMapRequest gmr = new GebetaMapRequest();
  List<ResponseData> rs = await gmr.matrix(points, apiKey);
  for (int i = 0; i < rs.length; i++) {
    print(rs[i].message);
    print(rs[i].status);
    print(rs[i].totalDistance);
    print(rs[i].path);
    print("(------------------------------------------)");
  }
}

void callTss() async {
  var points = <Map>[];
  String apiKey = "";
  points.add({'lat': 9.005980058445639, 'lon': 38.785734616513466});
  points.add({'lat': 9.01166345564756, 'lon': 38.789008246478424});
  GebetaMapRequest gmr = new GebetaMapRequest();
  ResponseData rds = await gmr.tss(points, apiKey);
  print('Response message: ${rds.message}');
  print('Response status: ${rds.status}');
  print('path: ${rds.path}');
  print('distance: ${rds.totalDistance}');
}

void callOneToMany() async {
  var directionStart = new Map();
  directionStart["lat"] = 9.022528936095531;
  directionStart["lon"] = 38.80400061607361;
  var points = <Map>[];
  String apiKey = "";
  points.add({'lat': 9.005980058445639, 'lon': 38.785734616513466});
  points.add({'lat': 9.01166345564756, 'lon': 38.789008246478424});
  GebetaMapRequest gmr = new GebetaMapRequest();
  List<ResponseData> rs = await gmr.OneToMany(directionStart, points, apiKey);

  for (int i = 0; i < rs.length; i++) {
    print(rs[i].message);
    print(rs[i].status);
    print(rs[i].totalDistance);
    print(rs[i].path);
    print("(------------------------------------------)");
  }
}

void callGeoCode() async {
  String name = "kotebe 02";
  String apiKey = "";
  GebetaMapRequest gmr = new GebetaMapRequest();
  ResponseData rs = await gmr.geocode(name, apiKey);
  print(rs.data);
}

void main(List<String> arguments) async {
  // making geocoding request
  callGeoCode();
}
